python/minetester/minetest_env.py

```diff
@@ -0,0 +1,505 @@
+        pygame.display.update()
+
+    def seed(self, seed: Optional[int] = None):
+        self._np_random = np.random.RandomState(seed or 0)
+
+    def reset(
+        self, seed: Optional[int] = None, options: Optional[Dict[str, Any]] = None
+    ):
+        self.seed(seed=seed)
+        if self.reset_world:
+            self._delete_world()
+            if self.reseed_on_reset:
+                self.world_seed = self._np_random.randint(np.iinfo(np.int64).max)
+        self._reset_minetest()
+        self._reset_zmq()
+        self._perform_client_handshake()
+
+        # Receive initial observation
+        logging.debug("Waiting for first obs...")
+        byte_obs = self.socket.recv()
+        (
+            obs,
+            _,
+            _,
+        ) = unpack_pb_obs(byte_obs)
+        self.last_obs = obs
+        logging.debug(f"Received first obs: {obs.shape}")
+        return obs, {}
+
+    def step(self, action: Dict[str, Any]):
+        assert (
+            self.client_process.poll() is None
+        ), f"Client process has terminated! Return code: {self.client_process.returncode}, logs in {self.log_dir}"
+
+        # Send action
+        if isinstance(action["MOUSE"], np.ndarray):
+            action["MOUSE"] = action["MOUSE"].tolist()
+        pb_action = pack_pb_action(action)
+        logging.debug(f"Sending action: {pb_action}")
+        self.socket.send(pb_action.to_bytes())
+
+        # TODO more robust check for whether a server/client is alive while receiving observations
+        if self.client_process is not None and self.client_process.poll() is not None:
+            return self.last_obs, 0.0, True, False, {}
+
+        # Receive observation
+        logging.debug("Waiting for obs...")
+        byte_obs = self.socket.recv()
+        next_obs, rew, done = unpack_pb_obs(byte_obs)
+
+        self.last_obs = next_obs
+        logging.debug(f"Received obs - {next_obs.shape}; reward - {rew}")
+
+        if self.render_mode == "human":
+            self._display_pygame()
+
+        return next_obs, rew, done, False, {}
+
+    def render(self):
+        if self.render_mode == "human":
+            # rendering happens during step, as per gymnasium API
+            return None
+        return self.last_obs
+
+    def close(self):
+        if self.socket is not None:
+            self.socket.close()
+        # TODO improve process termination
+        # i.e. don't kill, but close signal
+        if self.client_process is not None:
+            self.client_process.kill()
+        if self.reset_world:
+            self._delete_world()
+        if self.clean_config:
+            self._delete_config()
+
+    def _start_minetest_client(
+        self,
+        log_path: str,
+    ):
+        cmd = [
+            self.minetest_executable,
+            "--go",  # skip menu
+            "--config",
+            self.config_path,
+            "--remote-input",
+            f"127.0.0.1:{self.env_port}",
```

Comment:
sometimes you use 127.0.0.1 sometimes localhost. I suggest save the host:port to a member variable rather than saving self.env_port. Or does it need to be localhost in one usage and 127.0.0.1 at other usage? (I did find weird stuff with IPv6 and localhost which made me use 127.0.0.1 everywhere in the past I think)

Reply:
Yeah it's really random but ZMQ on some systems doesn't like localhost and needs 127.0.0.1, see https://stackoverflow.com/questions/6024003/why-doesnt-zeromq-work-on-localhost

127.0.0.1 is probably the way to go